### PR TITLE
Feature/dask

### PR DESCRIPTION
**Pull request by markymarkdavis**
This will be adding dask functionality by using the dask-ml pairwise distance functionality. There are tests and a notebook that are added to this to help show functionality.

### CODE REVIEW

**Review by hleerighter8 — COMMENTED** [review]
just some initial comments: looks like the first cell in the notebook has some import errors. secondly, I see that in the gam unit tests they don't seem to test the actual values, is that something you've tested in another place?

File: gam/gam.py [hunk 1]
```diff
@@ -54,7 +55,7 @@ class GAM:
     def __init__(
         self,
         k=2,
-        attributions_path="local_attributions.csv",
+        attributions_path=None,
         attributions=None,
```
[the remark this answers is not in the document]
Comment (markymarkdavis) [reply gam/gam.py:59]: @hleerighter8 We still want the ability to pass in a path to GAM, so `attributions_path` is fine. I'll need to update the docstring for `attributions` though, because we also do want the ability to pass in a dataframe or array, so I don't really want to change that to attributions_df since it could be either one of them.

File: tests/test_kendall_tau_distance.py [hunk 2]
```diff
@@ -22,15 +28,41 @@ def test_pairwise_distance_matrix():
     r1 = [0.05, 0.2, 0.7, 0.05]
     r2 = [0.23, 0.24, 0.26, 0.27]
     r3 = [0.22, 0.24, 0.26, 0.28]
-    rankings = [r1, r2, r3]
+    rankings = np.array([r1, r2, r3])
     D = pairwise_distance_matrix(rankings)
+
+    # Testing dask
+    client = Client(memory_limit="auto")
```
Comment (Lazea) [inline tests/test_kendall_tau_distance.py:35]: Don't need to set `auto` memory. Client's default is already `auto` I think.

File: tests/test_gam.py [hunk 3]
```diff
@@ -22,6 +26,32 @@ def test_read_df_or_list():
     att_arr = np.asarray(df.values.tolist())
     feat_labels_arr = np.asarray(df.columns.tolist())
 
+    client = Client(memory_limit="auto")
```
Comment (Lazea) [inline tests/test_gam.py:29]: No need to set to `auto` here.

File: tests/test_gam.py [hunk 4]
```diff
@@ -173,3 +203,17 @@ def test_plotting_2attributes():
     assert len(output) > 0
     for ofile in output:
         os.remove(ofile)
+
+def test_dask_vs_numpy():
+    client = Client(memory_limit="auto")
```
Comment (Lazea) [inline tests/test_gam.py:208]: No need for `auto` here.

File: tests/test_gam.py [hunk 5]
```diff
@@ -22,6 +26,32 @@ def test_read_df_or_list():
     att_arr = np.asarray(df.values.tolist())
     feat_labels_arr = np.asarray(df.columns.tolist())
 
+    client = Client(memory_limit="auto")
+    ddf = dd.read_csv("tests/test_attributes.csv")
```
Comment (Lazea) [inline tests/test_gam.py:30]: One thing to consider is that dask will not partition files on its own unless the data is large. You should try a `ddf = ddf.repartition(npartitions=4)` here to ensure the dask code is working with partitioned data.

File: tests/test_gam.py [hunk 6]
```diff
@@ -22,6 +26,32 @@ def test_read_df_or_list():
...
+    dask_att_arr = da.from_array(att_list)
+    dask_feat_labels_arr = da.from_array(feat_labels_list)
+
+    # Testing dask DataFrame
+    dask_df = gam.GAM(attributions=ddf)
```
Comment (Lazea) [inline tests/test_gam.py:35]: I would consider changing the name from `dask_df` to `g_ddf` otherwise this is confusing. Initially it made me think you are working with a dask dataframe and not the GAM object.

File: tests/test_gam.py [hunk 7]
```diff
@@ -173,3 +203,17 @@ def test_plotting_2attributes():
     assert len(output) > 0
     for ofile in output:
         os.remove(ofile)
+
+def test_dask_vs_numpy():
+    client = Client(memory_limit="auto")
+    ddf = dd.read_csv("tests/test_attributes.csv")
+    dask_df = gam.GAM(attributions=ddf)
```
Comment (Lazea) [inline tests/test_gam.py:210]: See my comment in the above test function about the `dask_df` name here.

File: tests/test_gam.py [hunk 8]
```diff
@@ -22,6 +26,32 @@ def test_read_df_or_list():
...
+    dask_df.generate()
+
+    assert hasattr(dask_df, "attributions")
+    assert dask_df.attributions.shape == (4, 3)
+
+    assert hasattr(dask_df, "feature_labels")
+    assert dask_df.feature_labels == ["a1", "a2", "a3"]
+
+    # Testing dask array
+    dask_list = gam.GAM(attributions=dask_att_arr, feature_labels=dask_feat_labels_arr)
```
Comment (Lazea) [inline tests/test_gam.py:45]: Same issue here. Try naming this to maybe `g_dask_list` or `g_da`.

File: gam/spearman_distance.py [hunk 9]
```diff
@@ -57,17 +60,21 @@ def spearman_squared_distance_legacy(r_1, r_2):
...
         rankings (numpy.array): Normalized Attributions
+        dask (boolean): whether or not to use dask's implementation
 
     Returns:
         [array[array]]: Spearman Distance Matrix
     """
-    return pairwise_distances(rankings, metric=spearman_squared_distance)
-
+    if dask:
+        D = dask_pairwise_distances(da.array(rankings), rankings, metric=spearman_squared_distance)
```
Comment (Lazea) [inline gam/spearman_distance.py:74]: I would consider instead allowing the user to pass in dask arrays directly and this function can then call on the appropriate distance function in response. This way the user has control over how the data is read in and partitioned.

File: gam/spearman_distance.py [hunk 10]
```diff
@@ -57,17 +60,21 @@ def spearman_squared_distance_legacy(r_1, r_2):
     return distance
 
 
-def pairwise_spearman_distance_matrix(rankings):
+def pairwise_spearman_distance_matrix(rankings, dask=False):
```
Comment (Lazea) [inline gam/spearman_distance.py:63]: See my comment below.

File: gam/kendall_tau_distance.py [hunk 11]
```diff
@@ -147,10 +147,15 @@ def mergeSortDistance(r1, r2):
...
-def pairwise_distance_matrix(rankings):
+def pairwise_distance_matrix(rankings, dask=False):
     from sklearn.metrics import pairwise_distances
-
-    D = pairwise_distances(rankings, metric=mergeSortDistance)
+    import numpy as np
+    from dask_ml.metrics.pairwise import pairwise_distances as dask_pairwise_distances
+    import dask.array as da
+    if dask:
+        D = dask_pairwise_distances(da.array(rankings), rankings, metric=mergeSortDistance)
```
Comment (Lazea) [inline gam/kendall_tau_distance.py:156]: Mentioned this in `spearman_disntance`, that you should consider letting users pass in dask arrays rather than creating the dask arrays yourself here.

File: gam/kendall_tau_distance.py [hunk 12]
```diff
@@ -147,10 +147,15 @@ def mergeSortDistance(r1, r2):
     return dist
 
 
-def pairwise_distance_matrix(rankings):
+def pairwise_distance_matrix(rankings, dask=False):
```
Comment (Lazea) [inline gam/kendall_tau_distance.py:150]: See my comment below.

File: gam/clustering.py [hunk 13]
```diff
@@ -4,13 +4,16 @@
...
+import dask.array as da
 import math
 import sys
 import time
 from copy import deepcopy
 
 import matplotlib.pyplot as plt
 import numpy as np
+from sklearn.metrics import pairwise_distances
+from dask_ml.metrics.pairwise import pairwise_distances as dask_pairwise_distances
```
Comment (Lazea) [inline gam/clustering.py:16]: The two `pairwise_distances` here will conflict. The import from `dask_ml` will override the `sklearn` import. Would be good to use `as` to name each pairwise for clarity and to avoid conflicts like this.

File: gam/kendall_tau_distance.py [hunk 14]
```diff
@@ -147,10 +147,15 @@ def mergeSortDistance(r1, r2):
...
     from sklearn.metrics import pairwise_distances
-
-    D = pairwise_distances(rankings, metric=mergeSortDistance)
+    import numpy as np
+    from dask_ml.metrics.pairwise import pairwise_distances as dask_pairwise_distances
+    import dask.array as da
```
Comment (Lazea) [inline gam/kendall_tau_distance.py:154]: This is technically an anti pattern, having imports nested in functions. Usually you would want them at the top of the file.

**Review by hleerighter8 — COMMENTED** [review]
looks good. I just have a few comments to maybe think about and @Lazea maybe you can jump in if either of these make sense.
1. line 178, seems like you are using numpy operations on attributions, which can be a dask array or numpy array. would using a dask array operation make sense? ie `da.abs(attributions).sum(axis=1, keepdims=True)` if attributions is a dask array, or even using `.map_blocks`?
2. from the notebook, it looks like `explanations` is a list of string and dask arrays, would consider converting to string and array so it's more readable to a user.

File: gam/spearman_distance.py [hunk 15]
```diff
@@ -7,6 +7,9 @@
 - add tests
 """
 
+from sklearn.metrics import pairwise_distances as sklearn_pairwise_distances
```
Comment (Lazea) [inline gam/spearman_distance.py:10]: This is a duiplicate import of the below import of `sklearn.metrics`

File: tests/test_kendall_tau_distance.py [hunk 16]
```diff
@@ -1,7 +1,14 @@
+from gam.kendall_tau_distance import ktau_weighted_distance
+from gam.kendall_tau_distance import mergeSortDistance
+
+from gam.kendall_tau_distance import pairwise_distance_matrix
```
Comment (Lazea) [inline tests/test_kendall_tau_distance.py:4]: Duplicate imports of line 7 below.

File: tests/test_kendall_tau_distance.py [hunk 17]
```diff
@@ -22,8 +29,15 @@ def test_pairwise_distance_matrix():
     r1 = [0.05, 0.2, 0.7, 0.05]
     r2 = [0.23, 0.24, 0.26, 0.27]
     r3 = [0.22, 0.24, 0.26, 0.28]
-    rankings = [r1, r2, r3]
+    rankings = np.array([r1, r2, r3])
     D = pairwise_distance_matrix(rankings)
+
+    # Testing dask
+    client = Client()
+    D_dask = pairwise_distance_matrix(da.from_array(rankings))
```
Comment (Lazea) [inline tests/test_kendall_tau_distance.py:37]: Does this function not return a dask array? If so you would need to do a compute or at least move the `client.close()` statement after the assert.

File: tests/test_kendall_tau_distance.py [hunk 18]
```diff
@@ -33,6 +47,25 @@ def test_pairwise_distance_matrix():
...
+
+    D = pairwise_distance_matrix(rankings)
+    # check symmetry, within floating point rounding margin
+    assert (D[0][1] - D[1][0]) < 1e-9
+    # check diagonal is zero
+    assert D[1][1] == 0
+    assert D[2][2] == 0
+    # distance between r2 and r3 is closer than r2 and r1
+    assert D[1][2] < D[1][0]
+    client.close()
```
Comment (Lazea) [inline tests/test_kendall_tau_distance.py:65]: This is testing the same thing as the function above but exclusively on dask but the above tests both dask and numpy. I say remove the dask pieces from the above function and just ahve dask tested in this function exclusively.

**Review by Lazea — CHANGES_REQUESTED** [review]
Made some further comments. Bit of a nitpick about cell 8 of the `MushroomClusterDask` notebook, all import statements should be kept up top of the file.

File: tests/test_spearman_distance.py [hunk 19]
```diff
@@ -30,8 +32,14 @@ def test_pairwise_distance_matrix():
     r2 = np.array([0.23, 0.24, 0.26, 0.27])
     r3 = np.array([0.22, 0.24, 0.26, 0.28])
 
-    rankings = [r1, r2, r3]
+    rankings = np.array([r1, r2, r3])
     D = pairwise_spearman_distance_matrix(rankings)
+    # Testing dask
+    client = Client()
+    D_dask = pairwise_spearman_distance_matrix(da.from_array(rankings))
+    client.close()
```
Comment (Lazea) [inline tests/test_spearman_distance.py:40]: Made a comment in `test_kendall_tau_distance.py ` about having a specific function for testing dask and would reccomend the same here. Remove the dask parts here and write a new `test_dask_pairwise_distance_matrix` function.

**Review by Lazea — CHANGES_REQUESTED** [review]
Left a few more comments. Will review the core GAM script tomorrow.

**Review by Lazea — DISMISSED** [review]
LGTM. Great work.

**Review by rreilly10 — APPROVED** [review]
Looks straightforward and clean - I am going to approve for now and then give @tazitoo and opportunity to take a look tomorrow and merge

**Review by Lazea — APPROVED** [review]
LGTM

**Review by hleerighter8 — APPROVED** [review]
lgtm

**Review by tazitoo — APPROVED** [review]
looks good to me
